File: futurejava/core/src/main/java/Perft.java
import org.javafish.board.BoardState;
import org.javafish.move.Move;
import org.javafish.move.MoveList;

import static org.javafish.app.Benchmark.GIGA;
import static org.javafish.board.BoardState.fromFen;
import static org.javafish.board.Fen.START_POS;

public class Perft {

    public static long printPerft(BoardState state, int depth) {
        final long startTime = System.nanoTime();
        long nodes = 0;
        MoveList moves = state.generateLegalMoves();
        for (Move move : moves){
            System.out.print(move.uci() + " ");
            BoardState newState = state.doMove(move);
            long move_nodes = perft(newState, depth - 1);
            nodes += move_nodes;
            System.out.println(move_nodes);
        }
        final long end_time = System.nanoTime();
        double time_taken = ((double)(end_time - startTime)) / GIGA;
        System.out.println("Nodes: " + nodes);
        System.out.println("Time: " + time_taken + "s");
        System.out.println("NPS: " + nodes/time_taken);
        return nodes;
    }

    public static long perft(BoardState state, int depth) {
        MoveList moves = state.generateLegalMoves();
        if (depth == 1) {
            return moves.size();
        }
        long nodes = 0;
        for (Move move : moves) {
            BoardState newState = state.doMove(move);
            nodes += perft(newState, depth - 1);
        }
        return nodes;
    }

    public static void main(String[] args) {
        printPerft(fromFen(START_POS), 7);
    }
}

File: ts-engine/src/Perft.ts
import {BoardState} from './BoardState';
import {Move} from './Move';

export class Perft {

  /**
   * Performs a perft search on the given board state up to the specified depth.
   * @param state The current board state.
   * @param depth The depth of the search.
   * @returns The number of nodes (legal moves) generated.
   */
  public static perft(state: BoardState, depth: number): number {
    const moves = state.generateLegalMoves();

    if (depth === 1) {
      return moves.length; // In TypeScript, use .length instead of .size()
    }

    let nodes = 0;

    moves.forEach((move: Move) => {
      const newBoardState = state.doMove(move);
      const count = Perft.perft(newBoardState, depth - 1);
      nodes += count;
    });

    return nodes;
  }

  /**
   * Generates a detailed perft result in a string format.
   * @param state The current board state.
   * @param depth The depth of the search.
   * @returns A formatted string with per-move breakdown and total nodes searched.
   */
  public static perftString(state: BoardState, depth: number): string {
    const startTime = performance.now(); // Start time measurement (milliseconds)
    let nodes = 0;
    const moveList = state.generateLegalMoves();

    moveList.forEach((move: Move) => {
      //process.stdout.write(move.toString() + " "); // Print the move in UCI format on the same line
      console.info(move.toString() + " "); // Print the move in UCI format on the same line
      let moveNodes = 0;

      const newBoardState = state.doMove(move);
      if (depth > 1) {
        moveNodes = Perft.perft(newBoardState, depth - 1);
      } else {
        moveNodes = 1; // At depth 1, each legal move counts as 1
      }

      nodes += moveNodes;
      console.log(moveNodes); // Print moveNodes count for the move
    });

    const endTime = performance.now(); // End time measurement
    const timeTaken = (endTime - startTime) / 1000; // Convert to seconds

    // console.log(`Nodes: ${nodes}`);
    // console.log(`Time: ${timeTaken.toFixed(3)}s`);
    // console.log(`NPS: ${(nodes / timeTaken).toFixed(0)}`);

    return `Nodes: ${nodes}\nTime: ${timeTaken.toFixed(3)}s\nNPS: ${(nodes / timeTaken).toFixed(0)}`;
  }

}
